### ISSUE TITLE
Why these branches are explored rather than pruned vs. the opposite way?

### ISSUE THREAD

**Issue by emadboctorx** [optuna/optuna#2808]
I've been trying to optimize hyperparameters of a reinforcement learning agent. Despite following the instructions in the [documentation](https://optuna.readthedocs.io/en/v1.0.0/tutorial/pruning.html) and doing the following:

    while True:
        self.train_step()
        if trial:
            current_reward = np.mean(self.total_rewards)  # intermediate reward
            self.trial.report(current_reward, self.steps)
            if self.trial.should_prune():
                raise optuna.exceptions.TrialPruned()
        if self.training_done():
            break

I occasionally end up with very poor end result trials, that should be pruned while they don't and waste a lot of time completing.

Here's the study [ppo.tar.gz](https://github.com/optuna/optuna/files/6821621/ppo.tar.gz) which has ~= 200 trials. Use the following to read the study:

    >>> import optuna
    >>> study = optuna.load_study('ppo-trials-with-pruning', storage='sqlite:///ppo.db')

The best trial # 83, has a reward value of 10.67, and note that I specify `direction='maximize'` to maximize the reward when I call `study.optimize()`. Why trial # 213 in the example below completes? shouldn't such trials be pruned? If you examined all the trials, you'll find many other similar examples that come after trial # 83. I didn't get into how the pruning logic works, however I don't think this behavior is intended.

    211 number                                         211
    value                                         1.04
    datetime_start          2021-07-15 08:13:13.901519
    datetime_complete       2021-07-15 08:15:18.766776
    duration                    0 days 00:02:04.865257
    params_clip_norm                           0.69256
    params_entropy_coef                       0.000005
    params_epsilon                            0.086382
    params_grad_norm                          1.785552
    params_lam                                0.952748
    params_learning_rate                      0.005663
    params_n_envs                                   16
    params_n_steps                                   4
    state                                       PRUNED
    Name: 211, dtype: object
    212 number                                         212
    value                                         0.84
    datetime_start          2021-07-15 08:15:25.353898
    datetime_complete       2021-07-15 08:17:28.525140
    duration                    0 days 00:02:03.171242
    params_clip_norm                          0.396523
    params_entropy_coef                       0.000002
    params_epsilon                            0.000003
    params_grad_norm                          1.606798
    params_lam                                 0.94495
    params_learning_rate                       0.00474
    params_n_envs                                   16
    params_n_steps                                   4
    state                                       PRUNED
    Name: 212, dtype: object
    213 number                                         213
    value                                         0.14
    datetime_start          2021-07-15 08:17:35.115506
    datetime_complete       2021-07-15 08:27:54.118670
    duration                    0 days 00:10:19.003164
    params_clip_norm                          0.485843
    params_entropy_coef                       0.000006
    params_epsilon                            0.000826
    params_grad_norm                           4.55236
    params_lam                                0.969098
    params_learning_rate                      0.007553
    params_n_envs                                   16
    params_n_steps                                   8
    state                                     COMPLETE

And later I got the exact opposite: a trial with a reward of 7.59 is pruned:

    number                                         232
    value                                         4.51
    datetime_start          2021-07-15 09:03:25.761148
    datetime_complete       2021-07-15 09:11:38.920445
    duration                    0 days 00:08:13.159297
    params_clip_norm                            0.6573
    params_entropy_coef                    0.000654174
    params_epsilon                         2.79855e-06
    params_grad_norm                           3.75322
    params_lam                                0.980803
    params_learning_rate                    0.00755071
    params_n_envs                                   16
    params_n_steps                                  64
    state                                     COMPLETE
    Name: 232, dtype: object
    
    number                                         233
    value                                         7.59
    datetime_start          2021-07-15 09:11:45.789277
    datetime_complete       2021-07-15 09:13:40.521504
    duration                    0 days 00:01:54.732227
    params_clip_norm                          0.660862
    params_entropy_coef                      0.0179951
    params_epsilon                         3.79014e-06
    params_grad_norm                           3.78325
    params_lam                                0.981147
    params_learning_rate                    0.00747825
    params_n_envs                                   16
    params_n_steps                                  64
    state                                       PRUNED
    Name: 233, dtype: object

### Environment

- Optuna version: 2.8.0
- Python version: 3.7
- OS: Ubuntu 18 (google colab)
- (Optional) Other libraries and their versions:

### Error messages, stack traces, or logs

```
# error messages, stack traces, or logs
```

### Steps to reproduce

1.
2.
3.

### Reproducible examples

```python
# python code
```

## Note to the questioner

If you are more comfortable with [Stack Overflow], you may consider posting your question with the "optuna" tag there instead.
Alternatively, for issues that would benefit from more of an interactive session with the developers,
you may refer to the [optuna/optuna chat] on Gitter.

[Stack Overflow]: https://stackoverflow.com/questions/tagged/optuna
[optuna/optuna chat]: https://gitter.im/optuna/optuna

**Comment by emadboctorx** [comment]
It was my mistake, I was reporting intermediate rewards with inconsistent number of steps. The problem is now gone after passing consistent steps during each trial.
